src/adapter/src/client.rs

```diff
@@ -747,3 +750,65 @@ impl Timeout {
         }
     }
 }
+
+/// A wrapper around an UnboundedReceiver of PeekResponseUnary that records when it sees the
+/// first row data in the given histogram
+#[derive(Debug)]
+pub struct RecordFirstRowStream {
+    pub rows: UnboundedReceiver<PeekResponseUnary>,
+    pub execute_started: Instant,
+    pub time_to_first_row_seconds: Histogram,
+    saw_rows: bool,
+}
+
+impl Future for RecordFirstRowStream {
```

Comment:
Generally `Future`s are not supposed to be `poll`-ed after they have returned `Ready`, see [docs](https://doc.rust-lang.org/std/future/trait.Future.html#return-value). Instead what we should implement here is [`Stream`](https://docs.rs/futures/latest/futures/stream/trait.Stream.html).

Although, since we have the `async fn recv()`, I'm not sure if we need to manually implement these traits? Could we do something like:

```
async fn recv(&mut self) -> Option<PeekResponseUnary> {
  let msg = self.rows.next().await;
  match &msg {
    // ...check saw_rows
  }
  msg
}
```

Reply:
ah thanks, great suggestion